hl_extractor/hl_calc.py

```diff
@@ -131,85 +111,56 @@ def create_profile(in_file, out_file, sha1):
 def get_build_sha1(binary):
     """Calculate the SHA1 of the binary we're using."""
     try:
-        f = open(binary, "r")
-        bin = f.read()
-        f.close()
+        with open(binary, "rb") as fp:
+            contents = fp.read()
     except IOError as e:
         print("Cannot calculate the SHA1 of the high-level extractor binary: %s" % e)
         sys.exit(-1)
 
-    return sha1(bin).hexdigest()
+    return hashlib.sha1(contents).hexdigest()
+
+
+def save_hl(mbid, rowid, hl_data, build_sha1):
+    try:
+        jdata = json.loads(hl_data)
+    except ValueError:
+        print("error %s: Cannot parse result document" % mbid)
+        print(hl_data)
+        jdata = {}
 
+    db.data.write_high_level(mbid, rowid, jdata, build_sha1)
 
-def main(num_threads):
+
+def cf_main(num_threads):
     print("High-level extractor daemon starting with %d threads" % num_threads)
-    sys.stdout.flush()
     build_sha1 = get_build_sha1(HIGH_LEVEL_EXTRACTOR_BINARY)
     create_profile(PROFILE_CONF_TEMPLATE, PROFILE_CONF, build_sha1)
 
     num_processed = 0
 
-    pool = {}
-    docs = []
-    while True:
-        # Check to see if we need more database rows
-        if len(docs) == 0:
-            # Fetch more rows from the DB
-            docs = db.data.get_unprocessed_highlevel_documents()
-
-            # We will fetch some rows that are already in progress. Remove those.
-            in_progress = pool.keys()
-            filtered = []
-            for mbid, doc, id in docs:
-                if mbid not in in_progress:
-                    filtered.append((mbid, doc, id))
-            docs = filtered
-
-        if len(docs):
-            # Start one document
-            mbid, doc, id = docs.pop()
-            th = HighLevel(mbid, doc, id)
-            th.start()
-            print("start %s" % mbid)
-            sys.stdout.flush()
-            pool[mbid] = th
-
-        # If we're at max threads, wait for one to complete
+    with concurrent.futures.ThreadPoolExecutor(max_workers=num_threads) as executor:
         while True:
-            if len(pool) == 0 and len(docs) == 0:
+            docs = db.data.get_unprocessed_highlevel_documents(DOCUMENTS_PER_QUERY)
+            print("Got %s documents" % len(docs))
+
+            future_to_llid = {}
+            for mbid, ll_json, rowid in docs:
+                future_to_llid[executor.submit(process_mbid, mbid, ll_json)] = (rowid, mbid)
+
+            for future in concurrent.futures.as_completed(future_to_llid):
+                rowid, mbid = future_to_llid[future]
+                try:
+                    hl_data = future.result()
+                except Exception as exc:
+                    print('%r generated an exception: %s' % (mbid, exc))
+                else:
+                    num_processed += 1
+                    save_hl(mbid, rowid, hl_data, build_sha1)
```

Comment:
Should we do some exception handling here too?  In case the db times out or something?

Reply:
yeah, we could move it into the try block and then have a specific except block to catch these errors. Generally, we haven't been catching db errors, but it's probably a good idea